File: c++/014/014.cpp
#include <iostream>

unsigned long long collatz(unsigned long long n) {
  if (n <= 1) {
    return 1;
  } else {
    if (n % 2 == 0) {
      return n / 2;
    } else {
      return (3 * n) + 1;
    }
  }
}

int main() {
  unsigned long long maxLength = 1;

  unsigned long long maxStart = 1;

  for (unsigned long long start = 1; start < 1000000; start += 1) {
    unsigned long long length = 1;

    unsigned long long n = start;

    while (n > 1) {
      n = collatz(n);

      length += 1;
    }

    if (length > maxLength) {
      maxLength = length;

      maxStart = start;
    }
  }

  std::cout << maxStart << std::endl;

  return 0;
}

File: rust/014/014.rs
fn collatz(n: u64) -> u64 {
    if n <= 1 {
        return 1;
    } else {
        if n % 2 == 0 {
            return n / 2;
        } else {
            return (3 * n) + 1;
        }
    }
}

fn main () {
    let mut max_length: u64 = 1;

    let mut max_start: u64 = 1;

    for start in 1..1000000 {
        let mut length: u64 = 1;

        let mut n: u64 = start;

        while n > 1 {
            n = collatz(n);

            length = length + 1;
        }

        if length > max_length {
            max_length = length;

            max_start = start;
        }
    }

    println!("{}", max_start);
}
